### DISCUSSION TITLE
issues two body dynamic bullet high velocity

### DISCUSSION THREAD

**Discussion by anhtu1995ok** [dyn4j/dyn4j#170] [General]
When the bullet collides it is frozen for a while

**Comment by NordelDev** [comment]
Can you provide a video or something like that? Then you can maybe receive better answers instead of wild guesses :-)

**Comment by dlwaters** [comment]
I'll post a sandbox demonstrating this issue. I've trying to isolate the
problem a little more before posting.

If ccd is on and velocity is above a certain (unknown) threshold, some
collisions are briefly sticky. Other bodies continue to move on screen but
there is a sometimes a sticky-collision that lasts a few steps.

It seems to happen more frequently the closer to a direct hit, not just a
glancing blow.

**Comment by dlwaters** [comment]
I think I have a lead on this. 

In ccd mode, as bodies move closer, there is a section in World.java that is supposed to nudge two bodies to within collision distance d. 

Starting at line 1046 of World.java.

		// make sure the time of impact is not null
		if (minToi != null) {
			// get the time of impact info
			double t = minToi.getTime();
			
			// move the dynamic body to the time of impact
			body1.transform0.lerp(body1.getTransform(), t, body1.getTransform());
			// check if the other body is dynamic
			if (minBody.isDynamic()) {
				// if the other body is dynamic then interpolate its transform also
				minBody.transform0.lerp(minBody.getTransform(), t, minBody.getTransform());
			}
			// this should bring the bodies within d distance from one another
			// we need to move the bodies more so that they are in collision
			// so that on the next time step they are solved by the discrete
			// collision detector
			
			// performs position correction on the body/bodies so that they are
			// in collision and will be detected in the next time step
			this.timeOfImpactSolver.solve(body1, minBody, minToi, this.settings);
			
			// this method does not conserve time
		}		// make sure the time of impact is not null
		if (minToi != null) {
			// get the time of impact info
			double t = minToi.getTime();
			
			// move the dynamic body to the time of impact
			body1.transform0.lerp(body1.getTransform(), t, body1.getTransform());
			// check if the other body is dynamic
			if (minBody.isDynamic()) {
				// if the other body is dynamic then interpolate its transform also
				minBody.transform0.lerp(minBody.getTransform(), t, minBody.getTransform());
			}
			// this should bring the bodies within d distance from one another
			// we need to move the bodies more so that they are in collision
			// so that on the next time step they are solved by the discrete
			// collision detector
			
			// performs position correction on the body/bodies so that they are
			// in collision and will be detected in the next time step
			this.timeOfImpactSolver.solve(body1, minBody, minToi, this.settings);
			
			// this method does not conserve time
		}

Below is a Sandbox file that creates two large circle bodies and one small.  The two large bodies are used to illustrate the hesitation, the small body is used to show that that small body keeps moving during the "hesitation" by the two colliding larger bodies.  

In order to see the issue, load this file into the sandbox (click snapshot to save the state, you'll need it in a minute) and then press <start simulation> to run it once to see how the hesitation seems to get worse after each new collision between the two large circular bodies. 

Stop the simulation, reset it using snapshot and then zoom in really tight on the origin to see what happens when the two large circluar bodies collide.  Use the "perform one simulation step" to step until the first collision occurs. 

[ccd_test_5.zip](https://github.com/dyn4j/dyn4j/files/5519326/ccd_test_5.zip)

What I discovered is that each time these large bodies get within a certain distance, each "step" moves them only a tiny bit forward, not quite enough to initiate contact:

			// this should bring the bodies within d distance from one another
			// we need to move the bodies more so that they are in collision
			// so that on the next time step they are solved by the discrete
			// collision detector

Worse, each time these two entities approach each other, the number of micro-steps increases rapidly, something like 3, 4, 12, 32 steps.  (That's not exact, it's tricky to count.)  That's why if you run the sandbox simulation you will notice that the "sticky-time" increases. 

What happens during this stickiness is the body "slows down" while it takes multiple micro-steps and then -- after collision finally occurs -- the bodies "jump apart" to put them in the position they should be in after the amount of time it took to collide. 

At least I think that is what is happening.  I've only been using dyn4j for a little over a week, so its entirely possible I'm totally off the mark.  I was researching this for several days before the issue was posted, so I did a have a head start. 

But, I don't know to fix it.  

My wild guesses: 

1) One of the solvers is returning a movement distance for each large circle-body that is too small to get within collision distance, so it takes several passes through this method to reach the required minimum distance. 

2) For some reason that solver is checking against the wrong "minimum distance" which is making the steps too small. 

Hope this helps.

**Comment by wnbittle** [comment]
@anhtu1995ok I would need more details than this to troubleshoot exactly - a video would be most helpful.

If you are referring to jittery movement during collision, this is a result of time-loss and the way CCD is handled in dyn4j.  The CCD solver detects collisions _**after**_ they've been missed.  If a missed collision is detected, the bodies are backtracked to the collision point.  This backtracking has the effect of losing time - the time between the missed collision and where it was before we detected the missed collision.  The effect on simulation is somewhat of a jitter/delay effect.

**Comment by dlwaters** [comment]
Here's a video illustrating my concerns.  

https://youtu.be/gYjobEXkQxY

I keep coming back to this comment in the code:

		// this should bring the bodies within d distance from one another
		// we need to move the bodies more so that they are in collision
		// so that on the next time step they are solved by the discrete
		// collision detector

When CCD is set to NONE, the simulation works fine.  When set to ALL, it's as if the "we need to move bodies so that they are in collision" isn't quite happening.  When CCD is set to ALL, each "step" moves the bodies closer but by an every decreasing amount until some threshold is broken and then the motion catches up to "current time" with a large jump. 

I've been trying for days to figure out exactly why this happens.  I suspect something related to this in World:

(World 924) 
solveTOI(Body body1, List<TimeOfImpactListener> listeners)

The lerp and "this.timeOfImpactSolver.solve don't seem to be moving the bodies into contact. 

(World 1051)
			// move the dynamic body to the time of impact
			body1.transform0.lerp(body1.getTransform(), t, body1.getTransform());
			// check if the other body is dynamic
			if (minBody.isDynamic()) {
				// if the other body is dynamic then interpolate its transform also
				minBody.transform0.lerp(minBody.getTransform(), t, minBody.getTransform());
			}
			// this should bring the bodies within d distance from one another
			// we need to move the bodies more so that they are in collision
			// so that on the next time step they are solved by the discrete
			// collision detector
			
			// performs position correction on the body/bodies so that they are
			// in collision and will be detected in the next time step
			this.timeOfImpactSolver.solve(body1, minBody, minToi, this.settings);

Raising the linear tolerance from .0005 to a larger number helps but this still feels like a bug, not a feature.  I plan on working with several thousand bullet-bodies (almost like an ideal gas simulation with actual collisions) and so far dyn4j has been great ... but that hesitation is really distracting. 

Let me know if there's anything I can do to clarify or help track this down.

**Comment by dlwaters** [comment]
Okay, I now realize my concerns may be related to what is known as the "motion clamping problem."

Motion clamping problem

https://digitalrune.github.io/DigitalRune-Documentation/html/138fc8fe-c536-40e0-af6b-0fb7e8eb9623.htm
This method is simple and works well in most cases, but it has a clear disadvantage: A part of the bodies' motion is simply discarded (the body loses time). Since CCD is used for fast bodies the motion clamping is mostly not visible – unless the motion is repeatedly clamped in several frames.

I'll dig some more now that I understand that at some level this is a "feature" and not a bug.

**Comment by wnbittle** [comment]
@dlwaters the link provided is a good description of the problem.  I would only add to that description that fast rotation, as in your example, makes things even harder and that the size, in addition to the speed, of the body matter.

That said, it may not explain everything you are seeing in dyn4j.  One problem I noticed while working on #121 and #122 is that the order of evaluation for CCD maybe incorrect - still testing this theory.  Another "problem" is that the default CCD mode is ALL, meaning all dynamic vs static AND dynamic vs dynamic.  The default mode will likely change in the next release to `BULLETS_ONLY` to improve stability and performance in the most common cases.

**Comment by anhtu1995ok** [comment]
[Captures.zip](https://github.com/dyn4j/dyn4j/files/5643808/Captures.zip)
This is video issue

**Comment by wnbittle** [comment]
@anhtu1995ok thanks for the video.  This is expected behavior.  Depending on your scenario, you can try turning off CCD:
```
world.getSettings().setContinuousDetectionMode(ContinuousDetectionMode.NONE);
// or only enable it for for bodies flagged as body.setBullet(true);
//world.getSettings().setContinuousDetectionMode(ContinuousDetectionMode.BULLETS_ONLY);
```
This is only necessary if you have small and fast bodies.

**Comment by wnbittle** [comment]
> Here's a video illustrating my concerns.
> 
> https://youtu.be/gYjobEXkQxY
> 
@dlwaters do you have a Sandbox output of this scenario?  I want to do some testing/troubleshooting with the CCD enhancements that are coming.  The zip posted earlier has a different scenario (3 circles).

**Reply by dlwaters** [reply]
When you say sandbox output, are you asking for a java output of the simulation or something else?  I had to take a break from coding but I'd be happy to get whatever you need. (I'll go back to coding, so I'm still interested!)

**Reply by wnbittle** [reply]
Thanks!  Either the XML or the Java export of the simulation highlighted in the video.  I'd like to test with some changes I've made recently to see if they'll have any improvement on the behavior. No rush or anything.

**Reply by dlwaters** [reply]
I think this is the one you are looking for.  It has "ALL" for the bullets.  I included the NONE and BULLETS_ONLY version as XML as well.  Let me know if this isn't what you are looking for. 
[ccd_rectangle_5_b_9.zip](https://github.com/dyn4j/dyn4j/files/5901694/ccd_rectangle_5_b_9.zip)

**Reply by wnbittle** [reply]
Thanks!  I'm happy to report that the enhancements I'm working on have really helped in this scenario (nearly removing it completely), but there still remain situations where this behavior persists, though not as often or apparent.  This will be getting released in 4.1.0 which _should_ be soon.

**Comment by dlwaters** [comment]
Improvement is good! Looking forward to testing it out.

**Comment by wnbittle** [comment]
I've published a new version of the library which includes significant changes to the CCD.  I'm not going to claim all issues are resolved, but from my testing things are much better.

**Reply by dlwaters** [reply]
I tested this and saw the Island oddness, though I didn't know what was causing it.  
The Island thing affected both bullet and non-bullet simulations so I ignored it.  
Your latest 4.1.2 seems to work quite well.  I don't see any of the slow-down-hesitation anymore with bullets on.  I haven't done detailed testing but the performance with bullets=true and ALL was quite similar to without bullets with 2000 entities with springs.  I'm quite pleased!  
I'd run a test in the Sandbox but I'm guessing that'll need some updates before it can use 4.1.2.

**Reply by wnbittle** [reply]
Great to hear!  I also noticed the strange Island behavior as well, but it took a while before I found the root cause.  Thankfully it was something easy and things look correct again.

Regarding CCD, there's still some hesitation in some scenarios, but I think it's all due to time loss now (rather than those really long periods of hesitation).  With respect to performance, my testing was with 6000 static bodies and 100 dynamic bodies and it was 10x faster with 4.1.x vs. 4.0.0.  It's been long time since I focused on CCD so happy to hear the effort has had a very positive effect.

Sandbox needs a refresh for sure - it's still stuck on 3.x if I remember correctly.... there's just not enough time in the day to keep it up to date these days.

**Comment by dlwaters** [comment]
I'm looking forward to trying it out! Slow, incremental progress!

**Comment by dlwaters** [comment]
I figured as much with the sandbox. A quick look at the errors when I
swapped out for the new library (just for kicks) made me think it would be
a significant rebuild! Only so much we can do.
